docs/rfcs/700-config-features-and-defaultpairs-via-connection-string.md

```diff
@@ -0,0 +1,225 @@
+
+So a valid connection string containing default pairs could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&default_server_side_encryption=xxxx&default_strorage_class=xxxx`
+
+Also, we generate the following function for default pairs to support configure the default pair during initialization in the way of `WithXxx()`:
+
+Take `default_storage_class` for example:
+
+```go
+func WithDefaultStorageClass(v string) Pair {
+	reutrn Pair{
+		Key:   "default_storage_class",
+		Value: v,
+	}
+}
+```
+
+The connection string above is equivalent to:
+
+```go
+store, err := s3.NewStorage(
+	ps.WithCredential("hmac:xxxx:xxxx"),
+	ps.WithName("bucket_name"),
+	ps.WithWorkDir("/prefix"),
+    }), 
+    s3.WithDefaultServiceSideEncryption("xxxx"),
+    s3.WithDefaultStorageClass("xxxx"),
+)
+
+```
+
+### Implementation
+
+#### Features
+
+`*Features` are types defined in service.
+
+**Feature Pairs Registry**
+
+From services side:
+
+We generate pairs like `enable_virtual_dir` according to `features` in `service.toml`. And register the feature pairs referring to [GSP-90: Re-support Initialization Via Connection String].
+
+```go
+var serviceFeaturesPairMap = map[string]string{
+    // ...
+    "default_loose_pair": "bool",
+}
+
+var storageFeaturesPairMap = map[string]string{
+    // ...
+    "default_loose_pair":  "bool",
+    "default_virtual_dir": "bool",
+}
+
+func init() {
+    // insert service and storage feature maps into `pairMap` for registry
+	for k, v := range serviceFeaturesPairMap {
+        pairMap[k] = v
+    }
+    for k, v := range storageFeaturesPairMap {
+        pairMap[k] = v
+    }
+    services.RegisterSchema("<type>", pairMap)
+}
+
+```
+
+According to the current logic, `New*FromString` will first split connection string parse it into `ps []Pairs`, and finally call `New*(ty, ps)`. Then we can handle the default feature pairs to convert to `*Features` in `ParsePair*New()`.
+
+#### Default Pairs
+
+Default pairs are defined in services.
+
+From [go-storage] side:
+
+- Defaultable field can be added to `Pair` in `specs` to mark whether the pair could be set as default while initiate.
+- Defaultable string array can be added to `Op` in `specs` to store the list of defaultable pair's name.
+- Defaultable pair array can be added to `Functions` in `definitions` to store the list of defaultable pairs.
+
+Based on the above, we can record all the default pairs and the default pairs for a specific operation while parsing service toml.
+
+From services side:
+
+We can add new field `defaultable` for operations, in which service can specify pairs that can be set as default, like:
+
+```toml
+[namespace.storage.op.write]
+optional = ["content_md5", "content_type", "io_callback", "storage_class"]
+defaultable = ["storage_class"]
+
+[namespace.storage.op.create_append]
+optional = ["content_type", "storage_class"]
+defaultable = ["storage_class"]
+```
+
+**Default pairs registry**

```

Comment:
The same question here.

How about generating logic in `generated.go`?

Reply:
Got it.